rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp

```diff
@@ -169,6 +169,24 @@ void SequentialWriter::close()
   storage_factory_.reset();
 }
 
+void SequentialWriter::register_message_definition(const rosbag2_storage::MessageDefinition & message_definition) {
+  if (type_names_to_message_definitions_.find(message_definition.name) !=
+    type_names_to_message_definitions_.end()) {
+    // message definition already registered.
+    return;
+  }
+
+  if (!storage_) {
+    throw std::runtime_error(
+      "Bag is not open. Call open() before registering a message definition.");
+  }
+
+  storage_->register_message_definition(message_definition);
+
+  type_names_to_message_definitions_.insert(std::make_pair(message_definition.name, message_definition));
+  // TODO something with converters
+}
+
 void SequentialWriter::create_topic(const rosbag2_storage::TopicMetadata & topic_with_type)
```

Comment:
Hi @james-rms, Yes it is a bit messy in regards to the usage of the `topics_info_mutex_`.  
As far as I remember @adamdbrw Tried to protect access to the message's counter in `topics_names_to_info_`
https://github.com/ros2/rosbag2/blob/76665ad8b977721c059b95d307697b8424fed07d/rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp#L423-L428
which are calling from cache consumer as call back from separate thread.
https://github.com/ros2/rosbag2/blob/76665ad8b977721c059b95d307697b8424fed07d/rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp#L143-L145
Since those mesages counter increment was calling from inner thread they were not protected under the `writer::writer_mutex_` and we have externally exposed API for removing topic 
https://github.com/ros2/rosbag2/blob/76665ad8b977721c059b95d307697b8424fed07d/rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp#L210-L220
Considering the flow how we are using rosbag2_cpp API from rosbag2_transport layer, we don't need this `topics_info_mutex_`. Since we are not calling `remove_topic(..)` API if subscription with write(..)  method in callback has been successfully created. 
https://github.com/ros2/rosbag2/blob/76665ad8b977721c059b95d307697b8424fed07d/rosbag2_transport/src/rosbag2_transport/recorder.cpp#L343-L355

However since we expose `Writer::remove_topic(const rosbag2_storage::TopicMetadata & topic_with_type)` API in theory if someone will create their own recorder they could potentially call in valid subscription has been created. I believe that `topics_info_mutex_` was added for such case.

In your case as you mentioned above you don't have ability to access `type_names_to_message_definitions_` without  locking `writer::writer_mutex_`. 
I agree that you don't need additional mutex for it.